cola/widgets/startup.py

```diff
@@ -265,16 +336,397 @@ def handle_broken_repo(self, index):
             self.context.settings.remove_recent(repo['path'])
             self.context.settings.save()
 
-            item = self.bookmarks_model.item(index.row())
-            self.items.remove(item)
-            self.bookmarks_model.removeRow(index.row())
+            item = model.item(index.row())
+            items.remove(item)
+            model.removeRow(index.row())
+
+            other_model_row_count = other_model.rowCount()
+            for i in range(other_model_row_count):
+                other_index = other_model.index(i, 0)
+                other_item = other_model.itemFromIndex(other_index)
+
+                if (
+                    isinstance(other_item, PromptWidgetItem)
+                    and other_item.path == repo['path']
+                ):
+                    other_items.remove(other_item)
+                    other_model.removeRow(other_index.row())
+                    return
```

Comment:
This is pretty complex because it's trying to mutate the view to quickly remove this entry.

The original code was a little fast and loose in that it modified the models directly but it wasn't as complex. Would it be possible to eliminate this coupling?

This code also originally only dealt with the bookmarks list. It'd almost be better to not change that behavior for now so that it doesn't need to have this complicated "other model" coupling going on.

Update: after testing, I see a little bit of why this is being done -- the view is now split and it uses a different widget for the Recents and the Favorites.

I'll have to think about that.. in the past it was all presented as a single widget that contained both bookmarks and recent items. The rationale is that a single widget is easier to use and understand, even if it is a little overloaded under the hood.

With two widgets it means there are two things to scroll and there's some redundancy across the lists because repos can be in both the Recents and Favorites, which is not quite as nice.

What do you think? Is it better to present them in a single view or is the screen real estate and complexity of two views worth it?

Reply:
I see. I had no such idea.
I splited into two views because of the reasons below. But doing that made many things complecated and might feels redundant.

> - View spliting
> I thought too.
> But a case like below could be possible. `bookmarks` and `recent` widgets in main UI display repository's name, and I thought splitiing the view is one of the solutions for clearfying which name should be displayed.
> And I also thought that displaying two views seems to be more usable.
> But after I read all comments, I though *view shouldn't be splited*.

So, changing the code to display only bookmarks list view might be good. Is it OK?